README.md
[![GitHub license](https://img.shields.io/github/license/palcarazm/bs-darkmode.svg)](https://github.com/palcarazm/bs-darkmode/blob/master/LICENSE)
[![Latest release](https://img.shields.io/github/package-json/v/palcarazm/bs-darkmode/v1.X.X?logo=github)](https://github.com/palcarazm/bs-darkmode/releases)
[![JSDelivr Badge](https://img.shields.io/jsdelivr/npm/hm/bs-darkmode?color=success&label=hits&logo=jsdelivr&logoColor=white)](https://www.jsdelivr.com/package/npm/bs-darkmode)
[![NPM Badge](https://img.shields.io/npm/dm/bs-darkmode?logo=npm&color=success)](https://www.npmjs.com/package/bs-darkmode)
[![Maintenance](https://img.shields.io/badge/maintained%3F-yes-success.svg)](https://github.com/palcarazm/bs-darkmode/graphs/contributors)
[![Funding](https://img.shields.io/badge/sponsor-30363D?style=flat&logo=GitHub-Sponsors&logoColor=#white)](https://github.com/sponsors/palcarazm)
[![Rate this package](https://badges.openbase.com/js/rating/bs-darkmode.svg?)](https://openbase.com/js/bs-darkmode?utm_source=embedded&amp;utm_medium=badge&amp;utm_campaign=rate-badge)

# Bootstrap Darkmode

**Bootstrap Darkmode** is a bootstrap plugin/widget that supports toggling between light and dark theme.

***

#### Library Distributions
Branch | Bootstrap Support | Last Release
---|---|---
[bs-darkmode v2.X.X](https://github.com/palcarazm/bs-darkmode/tree/v2.X.X)     | [![Bootstrap 5](https://img.shields.io/static/v1?label=bootstrap&message=v5.X.X&color=informational&logo=bootstrap&logoColor=white)](https://getbootstrap.com/docs/5.0) | [![Latest release](https://img.shields.io/github/package-json/v/palcarazm/bs-darkmode/v2.X.X?logo=github)](https://github.com/palcarazm/bs-darkmode/releases)
[bs-darkmode v1.X.X](https://github.com/palcarazm/bs-darkmode/tree/v1.X.X)     | [![Bootstrap 4](https://img.shields.io/static/v1?label=bootstrap&message=v4.X.X&color=informational&logo=bootstrap&logoColor=white)](https://getbootstrap.com/docs/4.0) | [![Latest release](https://img.shields.io/github/package-json/v/palcarazm/bs-darkmode/v1.X.X?logo=github)](https://github.com/palcarazm/bs-darkmode/releases)
# Demos
**Demos and API Docs:** https://palcarazm.github.io/bs-darkmode/  

# Related Bootstrap Puglins
<div align="center">
  <a href="https://github.com/palcarazm/bootstrap5-toggle" title="Boostrap Toggle"
    ><img
      src="https://github-readme-stats.vercel.app/api/pin/?username=palcarazm&repo=bootstrap5-toggle&border_radius=10&show_owner=true"
  /></a>
</div>

***

<!-- To update TOC run .\node_modules\.bin\doctoc README.md --github -->
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [Bootstrap Darkmode](#bootstrap-darkmode)
      - [Library Distributions](#library-distributions)
- [Demos](#demos)
- [Installation](#installation)
  - [CDN](#cdn)
    - [jQuery Interface](#jquery-interface)
    - [ECMAS Interface](#ecmas-interface)
  - [Download](#download)
  - [NPM](#npm)
  - [Yarn](#yarn)
- [Usage](#usage)
  - [Initialize With HTML](#initialize-with-html)
  - [Initialize With Code](#initialize-with-code)
- [API](#api)
  - [Options](#options)
  - [CSS Vars to change](#css-vars-to-change)
  - [Methods](#methods)
- [Events](#events)
  - [Event Propagation](#event-propagation)
  - [Stopping Event Propagation](#stopping-event-propagation)
- [Collaborators welcom!](#collaborators-welcom)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

**************************************************************************************************

# Installation

## CDN
[![JSDelivr Badge](https://img.shields.io/jsdelivr/npm/hm/bs-darkmode?color=success&label=hits&logo=jsdelivr&logoColor=white)](https://www.jsdelivr.com/package/npm/bs-darkmode)
### jQuery Interface
```html
<link href="https://cdn.jsdelivr.net/npm/bs-darkmode@1.2.0/css/bs-darkmode.min.css" rel="stylesheet">  
<script src="https://cdn.jsdelivr.net/npm/bs-darkmode@1.2.0/js/bs-darkmode.jquery.min.js"></script>
```

### ECMAS Interface
```html
<link href="https://cdn.jsdelivr.net/npm/bs-darkmode@1.2.0/css/bs-darkmode.min.css" rel="stylesheet">  
<script src="https://cdn.jsdelivr.net/npm/bs-darkmode@1.2.0/js/bs-darkmode.ecmas.min.js"></script>
```

## Download
[![Latest release](https://img.shields.io/github/package-json/v/palcarazm/bs-darkmode/v1.X.X?logo=github)](https://github.com/palcarazm/bs-darkmode/releases)

## NPM
[![NPM Badge](https://img.shields.io/npm/dm/bs-darkmode?logo=npm&color=success)](https://www.npmjs.com/package/bs-darkmode)
```ksh
npm install bs-darkmode@1.2.0
```

## Yarn
```ksh
yarn add bs-darkmode@1.2.0
```

# Usage

## Initialize With HTML
Simply add `data-plugin="bsdarkmode"` to automatically convert an element to a Bootstrap.

```html
<button class="btn btn-primary" data-plugin="bsdarkmode"></button>
```

## Initialize With Code
Toggles can also be initialized via JavaScript code.  

EX: Initialize id `darkmodeToggle` with a single line of JavaScript.
```html
<button class="btn btn-primary" id="darkmodeToggle"></button>
<script>
  $(function(){
    $('#darkmodeToggle').bsDarkmode();
  });
</script>
```

# API

## Options
* Options can be passed via data attributes or JavaScript
* For data attributes, append the option name to `data-` (ex: `data-state="light"`)

```html
<div id="api-example">
    <button class="btn btn-primary" data-plugin="bsdarkmode" data-state="dark" data-root="#api-example"
        data-lightvars='{"primary":"#FF0000","secondary":"#00FF00"}'
        data-darkvars='{"primary":"#0000FF","secondary":"#FFFF00"}'></button>
    <button class="btn btn-primary" id="darkmodeToggle"></button>
</div>
<script>
    $(function () {
        $('#darkmodeToggle').bsDarkmode({
            state: false,
            root:'#api-example',
            lightvars: '{"primary":"#FF0000","secondary":"#00FF00"}',
            darkvars: '{"primary":"#0000FF","secondary":"#FFFF00"}'
        });
    })
</script>
```

Name          |Type   |Default |Description                 
--------------|-------|--------|------------
`state`       |boolean|true    |Initial state (For data-state use `light` or `dark`). The user preferred color scheme dark is prioritary.
`root`        |string |":root" |Root element to apply CSS vars
`allowsCookie`|boolean|false   |Cookie authorization status
`lightlabel`  |html   |sun svg |Element inner HTML for light mode
`darklabel`   |html   |moon svg|Element inner HTML for dark mode
`lightvars`   |string |"{}"    |JSON object with CSS vars for light mode
`darkvars`    |string |"{}"    |JSON object with CSS vars for dark mode

## CSS Vars to change
The following CSS Vars that can be changed form `lightvars` and `darkvars` attributes.

Var             |Type     |Default Ligth|Default Dark|Description
----------------|---------|-------------|------------|-----------
`blue`          |HEX color|#0d6efd      |#3f6791     |Standard Bootstrap [color](https://getbootstrap.com/docs/4.0/getting-started/theming/#all-colors)
`indigo`        |HEX color|#6610f2      |#6610f2     |Standard Bootstrap [color](https://getbootstrap.com/docs/4.0/getting-started/theming/#all-colors)
`purple`        |HEX color|#6f42c1      |#6f42c1     |Standard Bootstrap [color](https://getbootstrap.com/docs/4.0/getting-started/theming/#all-colors)
`pink`          |HEX color|#d63384      |#e83e8c     |Standard Bootstrap [color](https://getbootstrap.com/docs/4.0/getting-started/theming/#all-colors)
`red`           |HEX color|#dc3545      |#e74c3c     |Standard Bootstrap [color](https://getbootstrap.com/docs/4.0/getting-started/theming/#all-colors)
`orange`        |HEX color|#fd7e14      |#fd7e14     |Standard Bootstrap [color](https://getbootstrap.com/docs/4.0/getting-started/theming/#all-colors)
`yellow`        |HEX color|#ffc107      |#f39c12     |Standard Bootstrap [color](https://getbootstrap.com/docs/4.0/getting-started/theming/#all-colors)
`green`         |HEX color|#198754      |#00bc8c     |Standard Bootstrap [color](https://getbootstrap.com/docs/4.0/getting-started/theming/#all-colors)
`teal`          |HEX color|#20c997      |#20c997     |Standard Bootstrap [color](https://getbootstrap.com/docs/4.0/getting-started/theming/#all-colors)
`cyan`          |HEX color|#0dcaf0      |#3498db     |Standard Bootstrap [color](https://getbootstrap.com/docs/4.0/getting-started/theming/#all-colors)
`gray`          |HEX color|#6c757d      |#ced4da     |Standard Bootstrap [gray color](https://getbootstrap.com/docs/4.0/getting-started/theming/#grays)
`gray-dark`     |HEX color|#343a40      |#e9ecef     |Standard Bootstrap [gray color](https://getbootstrap.com/docs/4.0/getting-started/theming/#grays)
`gray-100`      |HEX color|#f8f9fa      |#212529     |Standard Bootstrap [gray color](https://getbootstrap.com/docs/4.0/getting-started/theming/#grays)
`gray-200`      |HEX color|#e9ecef      |#343a40     |Standard Bootstrap [gray color](https://getbootstrap.com/docs/4.0/getting-started/theming/#grays)
`gray-300`      |HEX color|#dee2e6      |#495057     |Standard Bootstrap [gray color](https://getbootstrap.com/docs/4.0/getting-started/theming/#grays)
`gray-400`      |HEX color|#ced4da      |#6c757d     |Standard Bootstrap [gray color](https://getbootstrap.com/docs/4.0/getting-started/theming/#grays)
`gray-500`      |HEX color|#adb5bd      |#adb5bd     |Standard Bootstrap [gray color](https://getbootstrap.com/docs/4.0/getting-started/theming/#grays)
`gray-600`      |HEX color|#6c757d      |#ced4da     |Standard Bootstrap [gray color](https://getbootstrap.com/docs/4.0/getting-started/theming/#grays)
`gray-700`      |HEX color|#495057      |#dee2e6     |Standard Bootstrap [gray color](https://getbootstrap.com/docs/4.0/getting-started/theming/#grays)
`gray-800`      |HEX color|#343a40      |#e9ecef     |Standard Bootstrap [gray color](https://getbootstrap.com/docs/4.0/getting-started/theming/#grays)
`gray-900`      |HEX color|#212529      |#f8f9fa     |Standard Bootstrap [gray color](https://getbootstrap.com/docs/4.0/getting-started/theming/#grays)
`primary`       |HEX color|#0d6efd      |#3f6791     |Standard Bootstrap [theme color](https://getbootstrap.com/docs/4.0/getting-started/theming/#theme-colors)
`secondary`     |HEX color|#6c757d      |#ced4da     |Standard Bootstrap [theme color](https://getbootstrap.com/docs/4.0/getting-started/theming/#theme-colors)
`success`       |HEX color|#198754      |#00bc8c     |Standard Bootstrap [theme color](https://getbootstrap.com/docs/4.0/getting-started/theming/#theme-colors)
`info`          |HEX color|#0dcaf0      |#3498db     |Standard Bootstrap [theme color](https://getbootstrap.com/docs/4.0/getting-started/theming/#theme-colors)
`warning`       |HEX color|#ffc107      |#f39c12     |Standard Bootstrap [theme color](https://getbootstrap.com/docs/4.0/getting-started/theming/#theme-colors)
`danger`        |HEX color|#dc3545      |#e74c3c     |Standard Bootstrap [theme color](https://getbootstrap.com/docs/4.0/getting-started/theming/#theme-colors)
`light`         |HEX color|#f8f9fa      |#212529     |Standard Bootstrap [theme color](https://getbootstrap.com/docs/4.0/getting-started/theming/#theme-colors)
`dark`          |HEX color|#212529      |#f8f9fa     |Standard Bootstrap [theme color](https://getbootstrap.com/docs/4.0/getting-started/theming/#theme-colors)
`body-color`    |HEX color|#212529      |#f8f9fa     |Body color
`body-bg`       |HEX color|#fff         |#495057     |Body background color
`white`         |HEX color|#fff         |#000        |White color
`black`         |HEX color|#000         |#fff        |Black color

Theme colors, black and white are also exposed in HSL with `{color}-h`, `{color}-s` and `{color}-l`and in RGB with `{color}-rgb`.

## Methods
Methods can be used to control toggles directly.

```html
<button class="btn btn-primary" id="darkmodeToggle"></button>
```

Method               |Example                                                            |Description
---------------------|-------------------------------------------------------------------|------------
initialize           | `$('#darkmodeToggle').bsDarkmode()`                               |Initializes the darkmode plugin with options
light                | `$('#darkmodeToggle').bsDarkmode('light')`                        |Sets the darkmode toggle to 'light' state
dark                 | `$('#darkmodeToggle').bsDarkmode('dark')`                         |Sets the darkmode toggle to 'dark' state
toggle               | `$('#darkmodeToggle').bsDarkmode('toggle')`                       |Toggles the state of the darkmode toggle light/dark
setCookieAutorization| `$('#darkmodeToggle').bsDarkmode('setCookieAutorization', status)`|Sets the Cookie authorization status. Status can be `true` or `false`

# Events

## Event Propagation
Note All events are propagated to and from the element to the darkmode toggle.

You should listen to events from the HTML element directly rather than look for custom events.

```html
<button class="btn btn-primary" id="darkmodeToggle" data-plugin="bsdarkmode"></button>
<div id="console-event"></div>
<script>
  $(function() {
    $('#darkmodeToggle').change(function() {
      $('#console-event').html('Mode changed')
    })
  })
</script>
```

## Stopping Event Propagation
Passing `true` to the light, dark and toggle methods will enable the silent option to prevent the control from propagating the change event in cases where you want to update the controls light/dark state, but do not want to fire the onChange event.

```html
<button class="btn btn-primary" data-plugin="bsdarkmode" id="darkmodeToggle"></button>
<button class="btn btn-success" onclick="$('#darkmodeToggle').bsDarkmode('light',true)" >Light by API (silent)</button>
<button class="btn btn-success" onclick="$('#darkmodeToggle').bsDarkmode('dark',true)">Dark by API (silent)</button>
<button class="btn btn-warning" onclick="$('#darkmodeToggle').bsDarkmode('light')">Light by API (not silent)</button>
<button class="btn btn-warning" onclick="$('#darkmodeToggle').bsDarkmode('dark')">Dark by API (not silent)</button>
```

# Collaborators welcom!
- :sos: ¿Do you need some help? Open a issue in [GitHub help wanted](https://github.com/palcarazm/bs-darkmode/issues/new?assignees=&labels=help+wanted&template=help-wanted.md&title=%5BHELP%5D)
 - :bug: ¿Do you find a bug? Open a issue in [GitHub bug report](https://github.com/palcarazm/bs-darkmode/issues/new?assignees=&labels=bug&template=bug_report.md&title=%5BBUG%5D)
 - :bulb: ¿Do you have a great idea? Open a issue in [GitHub feature request](https://github.com/palcarazm/bs-darkmode/issues/new?assignees=&labels=enhancement&template=feature_request.md&title=%5BFEATURE%5D)
 - :computer: ¿Do you know how to fix a bug? Open a pull request in [GitHub pull repuest](https://github.com/palcarazm/bs-darkmode/compare).

[![GitHub Contributors](https://contrib.rocks/image?repo=palcarazm/bs-darkmode)](https://github.com/palcarazm/bs-darkmode/graphs/contributors)

¿Do you like the project? Give us a :star: in [GitHub](https://github.com/palcarazm/bs-darkmode).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased
### [Full Changelog](https://github.com/palcaraz/bs-darkmode/compare/v2.0.0...HEAD)

## [2.0.0](https://github.com/palcaraz/bs-darkmode/tree/v2.0.0) 2022-09-11
### Added
* feat: Support for Bootstrap 5 (#35) by @palcarazm

## [1.2.0](https://github.com/palcaraz/bs-darkmode/tree/v1.2.0) 2022-08-31
### Added
* feat: Set preferred color schema in a cookie (#8) by @palcarazm in https://github.com/palcarazm/bs-darkmode/pull/29
### Fixed
* fix: List Group style colors are not override (#31) by @palcarazm in https://github.com/palcarazm/bs-darkmode/pull/33
### [Full Changelog](https://github.com/palcarazm/bs-darkmode/compare/v1.1.1...v1.2.0)

## [1.1.1](https://github.com/palcaraz/bs-darkmode/tree/v1.1.1) 2022-08-22
### Fixed
* fix: Component render out of the root element by @palcarazm in https://github.com/palcarazm/bs-darkmode/pull/13
* fix: Jumbotron style colors are not override (#15) by @palcarazm in https://github.com/palcarazm/bs-darkmode/pull/17
* fix: Cards style colors are not override (#14) by @palcarazm in https://github.com/palcarazm/bs-darkmode/pull/18
* fix: Table styles colors are not override (#11) by @palcarazm in https://github.com/palcarazm/bs-darkmode/pull/21
* fix: Form and input group style colors are not override (#16) by @palcarazm in https://github.com/palcarazm/bs-darkmode/pull/24
* fix: Breadcrumb style colors are not override (#22) by @palcarazm in https://github.com/palcarazm/bs-darkmode/pull/25
* fix: Navbar and Navs style colors are not override (#23) by @palcarazm in https://github.com/palcarazm/bs-darkmode/pull/26
### [Full Changelog](https://github.com/palcarazm/bs-darkmode/compare/v1.1.0...v1.1.1)

## [1.1.0](https://github.com/palcaraz/bs-darkmode/tree/v1.1.0) 2022-08-13
### Added
* feat: Customize the root element to apply css vars (#3) by @palcarazm in https://github.com/palcarazm/bs-darkmode/pull/6
* feat : Default dark theme (#4) by @palcarazm in https://github.com/palcarazm/bs-darkmode/pull/9
* feat: Get user preferred color scheme (#7) by @palcarazm in https://github.com/palcarazm/bs-darkmode/pull/10
### Fixed
* fix: The change event is not fired when an action performed (#2) by @palcarazm in https://github.com/palcarazm/bs-darkmode/pull/5
* fix: Light CSS vars aren't applyed correctly by @palcarazm in https://github.com/palcarazm/bs-darkmode/pull/12

### [Full Changelog](https://github.com/palcarazm/bs-darkmode/compare/v1.0.0...v1.1.0)

## [1.0.0](https://github.com/palcaraz/bs-darkmode/tree/v1.0.0) 2022-08-10
- Original release

test/test-app.js
const ENV = $('#env-data')
/*const TESTCASES =[
    {name: 'status'         , code:'status'},
    {name: 'tristate'       , code:'tristate'},
    {name: 'size'           , code:'size'},
    {name: 'custom size'    , code:'custom-size'},
    {name: 'outline on'     , code:'outline-on'},
    {name: 'outline off'    , code:'outline-off'},
    {name: 'color on'       , code:'color-on'},
    {name: 'color off'      , code:'color-off'},
    {name: 'custom text'    , code:'custom-text'},
    {name: 'custom style'   , code:'custom-style'},
    {name: 'layout'         , code:'layout'},
    {name: 'API contructor' , code:'api-constructor'},
    {name: 'API methods' , code:'api-methods'}
];*/
function appStartup(test) {
    /*MAIN.html('');
    DESCRIPTION.html('');
    switch (test) {
    
        default:
            throw new DOMException('Unknown test case: '+ test, "NotSupportedError");
    }
    switch (INTERFACE) {
        case 'ECMAS':
            document.querySelectorAll('input[type=checkbox][data-toggle="toggle"]').forEach(function(ele) {
                ele.bootstrapToggle();
            });
            break;
        case 'JQUERY':
            $('input[data-toggle="toggle"]').bootstrapToggle();
            break;
    
        default:
            throw new DOMException('Unknown interface: '+ INTERFACE, "NotSupportedError");
    }
    
    setTimeout(function () {
        switch (test) {
    
        default:
            throw new DOMException('Unknown test case: '+ test, "NotSupportedError");
    }
    }, 500);*/
}
$( function () {
    $.getJSON('../package-lock.json', function(data) {
        Bootstrap = data.packages["node_modules/bootstrap"].version;
        plugin = data.version;
        ENV.html('');
        ENV.append(
            $("<div>").append($("<code>").html("Bootstrap v"+Bootstrap)),
            $("<div>").append($("<code>").html("bs-darkmode v"+plugin)),
            $("<div>").append($("<code>").html("Interface "+INTERFACE))
            )
    });

    $("section[data]").each(function (_index, section) {
        let request = new XMLHttpRequest();
        request.open("GET", $(section).attr("data"), true);
        request.send(null);
        request.onreadystatechange = function () {
          if (request.readyState === 4 && request.status === 200) {
            $(section).html(request.responseText);
          }
          if (request.readyState === 4 && request.status != 200) {
            $(section).html(
              $("<div></div>")
                .addClass("alert alert-warning")
                .attr("role", "alert")
                .html("Ouups! We can't load this section.")
            );
          }
        };
      });

    /*TESTCASES.forEach((testCase)=>{
        $('#test-selector').append(
            $('<button type="button">')
                .addClass("btn btn-secondary text-capitalize")
                .attr('data-test',testCase.code)
                .html(testCase.name)
        )
    });
    
    $('#test-selector button[data-test]').click(function() {
        appStartup($(this).attr('data-test'))
    });*/
});

js/bs-darkmode.ecmas.js
/* Copyright Notice
 * bs-darkmode v2.0.0
 * https://palcarazm.github.io/bs-darkmode
 * @author 2022 Pablo Alcaraz Martínez (https://github.com/palcarazm)
 * @funding GitHub Sponsors
 * @see https://github.com/sponsors/palcarazm
 * @license MIT
 * @see https://github.com/palcarazm/bs-darkmode/blob/master/LICENSE
 */


'use strict';

(function() {
	class Darkmode {
        DARKMODE_CLASS  = 'bs-darkmode-dark';
        LIGHTMODE_CLASS = 'bs-darkmode-light';
        COOKIE_NAME = 'bs-darkmode-color-scheme';

		constructor(element, options) {
			const DEFAULTS = {
                state: true,
                root: ':root',
                allowsCookie: false,
                lightlabel: '<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-brightness-high-fill" viewBox="0 0 16 16"><path d="M12 8a4 4 0 1 1-8 0 4 4 0 0 1 8 0zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"/></svg>',
                darklabel: '<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-moon-fill" viewBox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"/></svg>',
                lightvars:'{}',
                darkvars:'{}',
            }
			options = options || {};

			// A: Capture ref to HMTL element
			this.element = element;

			// B: Set options
            const allowsCookie = this.element.hasAttribute('allowsCookie') || options.allowsCookie || DEFAULTS.allowsCookie
            const COLORSCHEME = this.#getColorScheme(allowsCookie);
			this.options = {
				state: COLORSCHEME === null ? DEFAULTS.state : COLORSCHEME,
                root: this.element.getAttribute('data-root') || options.root || DEFAULTS.root,
                allowsCookie: allowsCookie,
                lightlabel: this.element.getAttribute('data-lightlabel') || options.lightlabel || DEFAULTS.lightlabel,
                darklabel: this.element.getAttribute('data-darklabel') || options.darklabel || DEFAULTS.darklabel,
                lightvars: this.element.getAttribute('data-lightvars') || options.lightvars || DEFAULTS.lightvars,
                darkvars: this.element.getAttribute('data-darkvars') || options.darkvars || DEFAULTS.darkvars,
			};

			// LAST: initialize
            this.#init();
		}

        #init() {
            // 1: Render darkmode
            this.#render();

            // 2: Add listener
            this.element.addEventListener('touchstart', (e)=>{this.#actionPerformed(e,this);});
            this.element.addEventListener('click', (e)=>{this.#actionPerformed(e,this);});
            if (window.matchMedia){
                window.matchMedia('(prefers-color-scheme: light)').addEventListener('change', (e) => {
                    if (this.options.state != e.matches){
                        this.#actionPerformed(e,this);
                    }
                });
            }

            // 3: Keep reference to this instance for subsequent calls via `getElementById().bsDarkmode()`
			this.element._bsDarkmode = this;
        }

        #render(){
            // 1: Add label
            this.element.innerHTML = this.options.state ? this.options.lightlabel : this.options.darklabel;

            // 2: Modify css vars
            let cssvars = JSON.parse(this.options.state ? this.options.lightvars : this.options.darkvars);
            if(!(typeof cssvars === 'object' && !Array.isArray(cssvars) && cssvars !== null)){
                throw new DOMException('Vars must be in a JSON Object.','TypeMismatchError');
            }
            const target = document.querySelector(this.options.root);
            target.removeAttribute("style");
            Object.entries(cssvars).forEach(([key, value]) => {
                target.style.setProperty("--bs-"+key,value);
            });

            // 3: Add HSL theme Colors
            const THEMECOLORS = ['primary', 'secondary', 'success', 'info', 'warning', 'danger', 'light', 'dark', 'white', 'black', 'body-color', 'body-bg'];
            let hsl, rgb;
            Object.entries(cssvars).forEach(([key, value]) => {
                if(THEMECOLORS.includes(key)){
                    rgb = this.#hexToRGB(value);
                    hsl = this.#hexToHSL(value);
                    target.style.setProperty("--bs-"+key+"-h",hsl.h+'');
                    target.style.setProperty("--bs-"+key+"-s",hsl.s+'%');
                    target.style.setProperty("--bs-"+key+"-l",hsl.l+'%');
                    target.style.setProperty("--bs-"+key+"-rgb",rgb.r+','+rgb.g+','+rgb.b);
                }
            });

            // 4: Add or remove Darkmode Class
            if (this.options.state) {
                target.classList.remove(this.DARKMODE_CLASS);
                target.classList.add(this.LIGHTMODE_CLASS);
            }else{
                target.classList.add(this.DARKMODE_CLASS);
                target.classList.remove(this.LIGHTMODE_CLASS);
            }
        }

        toggle = function (silent = false) {
            this.options.state = !this.options.state;
            this.#render();
            if (!silent) this.#trigger();
        }
    
        light = function (silent = false) {
            if (this.options.state) return;
            this.options.state = true;
            this.#render();
            if (!silent) this.#trigger();
        }
    
        dark = function (silent = false) {
            if (!this.options.state) return;
            this.options.state = false;
            this.#render();
            if (!silent) this.#trigger();
        }
    
        #trigger = function (silent) {
            if (!silent) this.element.dispatchEvent(new Event('change', { bubbles: true }));
        }

        setCookieAutorization = function (status) {
            this.options.allowsCookie = status;
        }

        /**
         * Get the color scheme to set
         * @param {Boolean} allowsCookie Cookie authorization status
         * @returns {Boolean} Color Scheme (light -> true / dark -> false)
         */
        #getColorScheme(allowsCookie){
            let state = null;

             // Cookie Preferred Scheme Dark
            if(allowsCookie){
                let cookie = this.#getCookie(this.COOKIE_NAME);
                if(cookie != null){
                    if (cookie === 'light') {
                        state = true;
                    }else if (cookie === 'dark'){
                        state =  false;
                    }
                }
            }

            // User Preferred Scheme Dark
            if(state === null){
                if (window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches){
                    state = false;
                }
            }

            // Element Data
            if(state === null){
                if (this.element.getAttribute('data-state') === 'light') {
                    state = true;
                }else if (this.element.getAttribute('data-state') === 'dark'){
                    state =  false;
                }
            }

            return state;    
        }

        /**
         * Trigger actions
         * @param {Event} e event
         * @param {Darkmode} target Darkmode
         * @private
         */
        #actionPerformed(e, target){
            target.toggle(false);
            if(target.options.allowsCookie){
                target.#setCookie(
                target.COOKIE_NAME,
                target.options.state ? "light" : "dark",
                0.25
                );
            }
            e.preventDefault();
        }

        /**
         * Convert Hex to HSL color space
         * @param {String} H Hex value '#rgb' or '#rrggbb'
         * @returns {Array} HSL value [h,s,l]
         */
        #hexToHSL(H) {
            // Convert hex to RGB first
            let rgb = this.#hexToRGB(H);
            let r = rgb.r;
            let g = rgb.g;
            let b = rgb.b;
            // Then to HSL
            r /= 255;
            g /= 255;
            b /= 255;
            let cmin = Math.min(r,g,b),
                cmax = Math.max(r,g,b),
                delta = cmax - cmin,
                h = 0,
                s = 0,
                l = 0;
            
            if (delta == 0)
                h = 0;
            else if (cmax == r)
                h = ((g - b) / delta) % 6;
            else if (cmax == g)
                h = (b - r) / delta + 2;
            else
                h = (r - g) / delta + 4;
            
            h = Math.round(h * 60);
            
            if (h < 0)
                h += 360;
            
            l = (cmax + cmin) / 2;
            s = delta == 0 ? 0 : delta / (1 - Math.abs(2 * l - 1));
            s = +(s * 100).toFixed(1);
            l = +(l * 100).toFixed(1);
            
            return {h:h,s:s,l:l};
        }

        /**
         * Convert Hex to RGB color space
         * @param {String} H Hex value '#rgb' or '#rrggbb'
         * @returns {Array} RGB value [r,g,b]
         */
        #hexToRGB(H) {
            let r = 0, g = 0, b = 0;
            if (H.length == 4) {
                r = parseInt("0x" + H[1] + H[1]);
                g = parseInt("0x" + H[2] + H[2]);
                b = parseInt("0x" + H[3] + H[3]);
            } else if (H.length == 7) {
                r = parseInt("0x" + H[1] + H[2]);
                g = parseInt("0x" + H[3] + H[4]);
                b = parseInt("0x" + H[5] + H[6]);
            }
            return {r:r,g:g,b:b};       
        }

        /**
         * Set a Cookie
         * @param {String} name Cookie Name
         * @param {String} value Cookie Value
         * @param {Number} days Days to expiration
         * @author Mandeep Janjua and Fakhruddin Ujjainwala
         * @see https://stackoverflow.com/a/24103596/17594569
         * @license CC BY-SA 4.0
         */
        #setCookie(name,value,days) {
            let expires = "";
            if (days) {
                let date = new Date();
                date.setTime(date.getTime() + (days*24*60*60*1000));
                expires = "; expires=" + date.toUTCString();
            }
            document.cookie = name + "=" + (value || "")  + expires + "; path=/";
        }

        /**
         * Get a Cookie
         * @param {String} name Cookie Name
         * @returns {String | null} Cookie Value
         * @author Srinivas Sabbani and Mohit Kumar Gupta
         * @see https://stackoverflow.com/a/4825695/17594569
         * @license CC BY-SA 4.0
         */
        #getCookie(name) {
            if (document.cookie.length > 0) {
                let start = document.cookie.indexOf(name + "=");
                if (start != -1) {
                    start = start + name.length + 1;
                    let end = document.cookie.indexOf(";", start);
                    if (end == -1) {
                        end = document.cookie.length;
                    }
                    return decodeURIComponent(document.cookie.substring(start, end));
                }
            }
            return null;
        }

        /**
         * Delete a Cookie
         * @param {String} name Cookie Name
         * @author Mandeep Janjua and Fakhruddin Ujjainwala
         * @see https://stackoverflow.com/a/24103596/17594569
         * @license CC BY-SA 4.0
         */
        #deleteCookie(name) {   
            document.cookie = name +'=; Path=/; Expires=Thu, 01 Jan 1970 00:00:00 UTC;';
        }
    }

	/**
	 * Add `bsDarkmode` prototype function to HTML Elements
	 * Enables execution when used with HTML - ex: `document.getElementById('toggle').bsDarkmode('light')`
	 */
	Element.prototype.bsDarkmode = function(options, args) {
		let _bsDarkmode = this._bsDarkmode || new Darkmode(this, options);

		if (options && typeof options === 'string') {
            switch (options.toLowerCase()) {
                case 'toggle':
                    _bsDarkmode.toggle(args);
                    break;
                case 'light':
                    _bsDarkmode.light(args);
                    break;
                case 'dark':
                    _bsDarkmode.dark(args);
                    break;
                case 'setCookieAutorization':
                    _bsDarkmode.setCookieAutorization(args);
                    break;
            
                default:
                    throw new DOMException("Bootstrap-Darkmode method '"+options+"' doesn't exist",'NotSupportedError');
            }
		}
	};

	/**
	 * Replace all `[data-plugin="bsdarkmode"]` inputs with "Bootstrap-Darkmode"
	 * Executes once page elements have rendered enabling script to be placed in `<head>`
	 */
	if (typeof window !== 'undefined')
		window.onload = function() {
			document.querySelectorAll('[data-plugin="bsdarkmode"]').forEach(function(ele) {
				ele.bsDarkmode();
			});
		};

	// Export library if possible
	if (typeof module !== 'undefined' && module.exports) {
		module.exports = Darkmode;
	}
})();